### PR TITLE
Feat/adaptor filter species

### PR DESCRIPTION
**Pull request by tjunge**
We've implemented a neutral `AdaptorFilter` which can be used to present a filtered view of an existing `StructureManager` based on arbitrary filter criteria. See `test_adaptor_filter.cc` for application examples, specifically `filter_1_test` and `filter_2_test`.

On the basis of these filters, we wrote `SpeciesManager` which explodes any `StructureManager` in a single iteration into `AdaptorFilters` for every existing species combination for clusters up a given `MaxOrder`. See `test_species_manager.cc` for application examples, specifically `atom_species_test` and `pair_species_test`. 

These filtered views are accessible by species combination (a simple `std::array<int, Order>`), see, e.g., line 101 in `test_species_manager.cc`.

### CODE REVIEW

**Comment by tjunge** [comment]
@ceriottm @felixmusil @max-veit 
Hi, could you please review this? we'll start a new feature dev from this merged with master, and it would be nice to get this integrated rather sooner than later in order to keep the number of branches low.

File: src/structure_managers/species_manager.hh [hunk 1]
```diff
@@ -30,34 +28,91 @@
...
+#include <tuple>
+#include <map>
+#include <memory>
 
 namespace rascal {
 
-  template <class NeighManager, int MaxLayer, int Layer = 0>
+  namespace internal {
+
+    namespace {
```
Comment (max-veit) [inline src/structure_managers/species_manager.hh]: Any specific reason we need an unnamed namespace here? This practice is discouraged for header files.

Comment (max-veit) [reply src/structure_managers/species_manager.hh]: Nevermind, just read the change to the coding convention. I guess it makes sense for implementation-heavy header files, but didn't we agree to be moving this sort of thing to separate "implementation header" (`_impl.hh`) files?

**Comment by felixmusil** [comment]
> @tjunge please have a look at the coding conventions changes I merged in; in particular, one of the new commits on master ([da476ea](https://github.com/cosmo-epfl/librascal/commit/da476ea204ffb4e9da7d4a3442acc71ad88dd244)) deletes some files (`structure_manager_chain.{cc,hh}`) that I'm not sure were meant to be deleted.

These files should be deleted.

**Comment by mastricker** [comment]
I think the discussion about the namespace is one for tomorrow.

As for the linting: I agree with Max - top level cfg file is the guideline.
This has to be automatic and fixed for everyone. We can and should not trade configs based on what we feel the code should look like. A added a point for discussion for tomorrow.

Lint target for test - also agree.

-- your points
4. Filter/Species business.
You understand correctly

5. Template argument MaxOrder is not necessarily the depth of possible combinations, but the desired.
You could in principle have a triplet-manager, but only want to have the species filtered up to the pair order.
Altough for other reasons than you state, I tend to agree that the naming is ambiguous and we should think about a different name.

6. I do not understand your point. The dereference of the structure manager was always there. We just had to change something with the cluster_index.

I still have to look at 2 and 7.

One other thing:
Do we allow reviewers to change code?
Is then someone else supposed to review the reviewers changed code?

**Comment by felixmusil** [comment]
About the filter options, I did not know there was such file and it turns out I have the same filter except that you added -build/header_guard which I am not sure why since it something we what to check for.
Maybe you could remove this filter from the CPPLINT.cfg file ? We can discuss this tomorrow too.

**Comment by max-veit** [comment]
> One other thing:
> Do we allow reviewers to change code?
> Is then someone else supposed to review the reviewers changed code?

The practice seems to differ between projects. Yes, sometimes it's faster for the reviewer to just do the change by themselves, especially if it's tiny things like formatting and conventions. My instinct in this case would be to wait from confirmation from the requester -- though I suppose that's not much faster than just saying "run cpplint on this".

Maybe we should make more extensive use of the "suggested/proposed changes" feature of github; anything small enough that you can type it up in the web interface can be done through there (and approved by the requester). For anything more complicated, I agree with Félix now: It should probably be done by the requester and re-reviewed when ready.

File: src/structure_managers/structure_manager.hh [hunk 2]
```diff
@@ -719,6 +781,20 @@ namespace rascal {
       return ClusterRef_t(*this, this->get_atom_indices(), cluster_indices);
     }
 
+    //! dereference: calculate cluster indices
```
[the remark this answers is not in the document]
Comment (mastricker) [reply src/structure_managers/structure_manager.hh]: Ah, ok I see. It is just the const-correct version of the existing cluster_indices container.
A couple more const correct getters were also added in this branch.

File: docs/source/coding-convention.rst [hunk 3]
```diff
@@ -286,7 +286,7 @@ Decision:
 Unnamed Namespaces and Static Variables
 ---------------------------------------
 
-When definitions in a ``.cc`` file do not need to be referenced outside that file, place them in an unnamed namespace or declare them static. Do not use either of these constructs in ``.hh`` files.
+When definitions do not need to be referenced outside that file, place them in an unnamed namespace or declare them static.
```
[the remark this answers is not in the document]
Comment (tjunge) [reply docs/source/coding-convention.rst]: After today's discussion, I'm suggesting to revert this to the original, and to keep discouraging anonymous namespaces in headers. Agreed?

**Comment by tjunge** [comment]
The linting and g++-5.4 compatibility are fixed, and I think the questions were answered. Can this now be merged, or are there more comments?

**Comment by felixmusil** [comment]
I still find linting error about the +1 indent of public (I think we converged on this standard if I recall) and a few others in adaptor_neighbour_list.hh:126 or the tests for instance. 
It seems to me that the filters for cpplint are a bit too many (see my list higher up in the chat). Also from our previous discussion we converged on using cpplint from pip (https://pypi.org/project/cpplint/) so you could maybe remove it from external and make the modifications to the 'make lint' accordingly now so that it is on master as soon as possible ?

So after the merge, we need to make a use case / tutorial to highlight the relationship between the FilterAdaptor and the SpeciesManager so that later on, one can easily code up something similar to  SpeciesManager.

**Comment by tjunge** [comment]
Ok, linting is now part of the CI.
For the future, I don't like that you insist on solving unrelated problems in single pull requests. i.e., the inclusion of the system linter has nothing to do with species management.

**Comment by felixmusil** [comment]
Good, this all seems. I just realized you did not answer on the bindings side. Are you planning to make some before merging or to merge and then continue working on the branch ?

@tjunge I am not sure to get you. I was suggesting getting this small thing into the master branch since you are likely to merge it in the following days. If you don't feel like it then don't follow my suggestion because it was not meant to make you unhappy.

**Comment by felixmusil** [comment]
Well...

**Comment by tjunge** [comment]
Good point about the bindings, I forgot to lint those. fixing it now

**Comment by mastricker** [comment]
I have added the bindings cpplint to the CI todo list

**Comment by felixmusil** [comment]
I was more referring to the fact that there are no bindings to the AdaptorFilter and SpeciesManager.

**Comment by tjunge** [comment]
Oh, now I understand. It's true that a smart python version of `AdaptorFilter` would be super useful for prototyping and debugging. `SpeciesManager` should be rather straight-forward.

**Pull request merged**